sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc

```diff
@@ -0,0 +1,644 @@
+# Matrix Programming Extension for DPC++: sycl_ext_oneapi_matrix
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+:dpcpp: pass:[DPC++]
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+Copyright (c) 2021-2021 Intel Corporation.  All rights reserved.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension is written against the SYCL 2020 revision 3 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+**_NOTE:_** _This document describes the current design and API for the matrix
+extension to {dpcpp}. This is an initial experimental version to try out functionality
+and performance, and **future versions of this API may change in ways that are incompatible with this experimental version**. The current implementation provides support of the matrix interface on Intel(R) Advanced Matrix Extensions (AMX) and DPAS_
+
+## Introduction
+This document presents an ongoing work towards defining a unified matrix interface. This interface is intended to unify different tensor hardware: Intel AMX in CPUs, DPAS in Intel GPUs, Habana Gaudi and Goya tensor and gemm cores, Nvidia TPUs, IBM Power MMA. All these hardware provide low-level intrinsics or assembly to access and perform matrix operations. The goal is to provide a unified interface that is portable but also benefit from the maximum performance these different hardware can offer.
+
+## Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_MATRIX` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[frame="none",options="header"]
+|======================
+|Value |Description
+|1     |Initial extension JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported 
+|2     |joint matrix type has a new `use` parameter. `layout` on matrix is unused. JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported 
```

Comment:
The `Layout` template parameter in this version seems confusing.  It seems like it is required to be `layout::unused` for most of the functions.  For example, `joint_matrix_load`, `joint_matrix_store`, and `joint_matrix_mad` all require the layout to be `layout::unused`.  Why not just remove that template parameter?  It seems like you can't set it to any other value.

Reply:
We cannot remove it because Jack requires it for the CUDA backend on the mad function. But, you are right, it should not be set to unused for these functions. This was based on a comment from Jack but I might have misunderstood him.  
@JackAKirk,  can you please comment here on how you will be able to set it for other values if it is unused?

Reply:
Yes on the mad function it is required for the cuda backend. If we want to have a single `joint_matrix_mad` function for all backends I suggested previously that we can however use the template parameter with a default value that is set layout::unused for the intel backend: I think this allows as far as the programmer is concerned the same interface as you would have if you delete the layout template parameter: as I've and implemented here: https://github.com/intel/llvm/pull/6657/files#diff-f71a436bdeda598b29caad471fa637a2844a12f38fe4e85b15b2ccb37bd09833R825

For the `joint_matrix_load` `joint_matrix_store` the CUDA backend uses two overloads of these functions: one for the c/accumulator matrix which matches exactly to the intel backend interface of these functions, and another overload that requires the template parameter of joint layout and doesnt have the regular parameter of joint layout: For the intel backend you don't need to worry about the second overload, but again, I think it would be nice if the first overload in the CUDA backend coincides with the Intel backend, so that only one function is used for both backends.

Reply:
The interfaces in this spec doc for `joint_matrix_load` and `joint_matrix_store` are already fine from my point of view. As described above there will can be an additional overload of `joint_matrix_load` wrapped in `#ifdef nvptx` that deals with the loads of use::a/use::b where the layout is required as a template parameter in the CUDA backend.

Reply:
> We cannot remove it because Jack requires it for the CUDA backend on the mad function. But, you are right, it should not be set to unused for these functions. This was based on a comment from Jack but I might have misunderstood him. @JackAKirk, can you please comment here on how you will be able to set it for other values if it is unused?

In Summary I think the Layout parameter should be set to unused by default in `joint_matrix_mad`, but it is not necessary at all in the other functions (which is actually as the current spec doc describes).